src/io/filewatcher.rs

```diff
@@ -4,120 +4,83 @@
-            if let Ok(modified) = file.metadata().unwrap().modified() {
-                if let Some(last_read) = self.last_read {
-                    if modified <= last_read {
-                        return None;
-                    }
-                } else {
-                    log::error!("No stored last mod time.");
-                }
-            } else {
-                log::error!("Unable to read modification time of watched file.");
-            }
-            let mut buff: Vec<u8> = Vec::new();
-            let _ = file.read_to_end(&mut buff);
+        // No new data
+        if meta.len() as usize == self.last_size || meta.len() == 0 {
+            return Ok(());
+        }
 
-            let mut reader = Cursor::new(buff);
-            reader.set_position(self.cpos);
+        // Get new file contents
+        let mut file =
+            FileWatcher::open_file(&self.file_path).context("Failed to reopen log file")?;
+        let mut buff: Vec<u8> = Vec::new();
+        let _ = file.read_to_end(&mut buff);
+        let mut start_idx = self.last_size;
 
-            if let Ok(mod_time) = file.metadata().unwrap().modified() {
-                Some(ReadMD(Some(mod_time), reader))
-            } else {
-                Some(ReadMD(None, reader)) // TODO: fix this grossness
-            }
-        } else {
-            None
+        // Reset if file has been remade (i.e. is shorter) and update state
+        if buff.len() < self.last_size {
+            start_idx = 0;
         }
-    }
-
-    fn read_lines(&mut self) {
-        if let Some(mut read_md) = self.seek_to_pos() {
-            self.last_read = read_md.0;
 
-            let mut data: Vec<u8> = Vec::new();
-            if let Ok(bytes_read) = read_md.1.read_to_end(&mut data) {
-                self.cpos += bytes_read as u64;
-            } else {
-                log::error!("Could not read file data!");
-                return;
-            }
+        self.last_size = buff.len();
 
-            let data_str = String::from_utf8_lossy(&data);
-            let mut lines: VecDeque<String> = VecDeque::new();
+        // Get strings
+        let data_str = String::from_utf8_lossy(&buff[start_idx..]);
+        self.lines_buf.extend(
+            data_str
+                .lines()
+                .filter(|x| !x.trim().is_empty())
+                .map(str::to_string),
+        );
 
-            self.lines_buf.extend(
-                data_str
-                    .lines()
-                    .filter(|x| !(x.trim().is_empty()))
-                    .map(str::to_string),
-            );
-        }
+        Ok(())
     }
 
-    pub fn get_line(&mut self) -> Option<String> {
+    /// Return the next
+    pub fn get_line(&mut self) -> Result<Option<String>> {
         if self.lines_buf.is_empty() {
-            self.read_lines();
+            self.read_new_file_lines()
+                .context("Failed to read new file lines.")?
```

Comment:
to be more user friendly, this should probably be a more generic error message like "failed to read file", im being pedantic

Reply:
I've reworded it a little bit, but am leaving in the detail because the context acts somewhat like a stack trace. If something does go wrong it would be more helpful for anybody the user might seek for help resolving an issue.